File: Container With Most Water/Container With Most Water.java
public class Solution {
    public int maxArea(int[] height) {
        int ret = 0;
        int left = 0;
        int right = height.length - 1;
        
        while(left < right){
            int width = right - left;
            int h = Math.min(height[left], height[right]);
            ret = Math.max(ret, width * h);
            if(height[left] < height[right])        ++left;
            else                                    --right;
        }
        
        return ret;
    }
}

File: Container With Most Water/Container With Most Water.cpp
class Solution {
public:

    int min(int left, int right){
        if(left < right)    return left;
        else                return right;
    }
    
    int max(int left, int right){
        if(left > right)    return left;
        else                return right;
    }

    int maxArea(vector<int> &height) {
        // 3 elements is the minimum requirement for forming a container
        if(height.size() < 2)       return 0;
        
        int leftIndex = 0;
        int rightIndex = height.size() - 1;
        
        int currVol = 0;
        int maxVol = 0;
        
        while(leftIndex < rightIndex){
            
            currVol = (rightIndex - leftIndex) * min(height[leftIndex], height[rightIndex]);
            maxVol = max(maxVol, currVol);
            
            if(height[leftIndex] <= height[rightIndex]){
                leftIndex++;
            }
            else{
                rightIndex--;
            }
        }
        
        return maxVol;
    }
};
